### PR TITLE
Add support for pulling from http location. 

### PR DESCRIPTION
**Pull request by plietar**
This adds two new location types, "http" and "packit", used to interact with outpack_server and Packit, respectively. The behaviour is based upon the R orderly2 implementation.

Authentication to Packit can be done by providing a Packit JWT, a GitHub PAT or interactively using an OAuth device flow against GitHub. For now, tokens obtained through interactive login are only cached in memory. I need to think about where and how tokens may be saved on disk.

There is no pushing support yet.

### CODE REVIEW

**Comment by plietar** [comment]
Thoughts on just dropping support for 3.8? It was released 5 years ago, and has just hit EOL last month.

`functools.cache` which is used to cache the tokens isn't available in that version.

**Comment by richfitz** [comment]
Dropping 3.8 is totally fine

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -36,6 +37,18 @@ jobs:
         run: |
           python -m pip install --upgrade pip
           pip install hatch
+
+      # Compiling outpack server from source takes a few minutes each time.
```
Comment (richfitz) [inline .github/workflows/test.yml:41]: nice!

File: pyproject.toml [hunk 2]
```diff
@@ -46,15 +49,16 @@ path = "src/pyorderly/__about__.py"
...
+  "pytest-unordered",
+  "pytest_mock",
+  "responses",
   "sphinx",
-  "sphinx-rtd-theme",
-  "myst-parser",
+  "sphinx-autoapi",
   "sphinx-copybutton",
-  "sphinx-autoapi"
+  "sphinx-rtd-theme",
```
Comment (richfitz) [inline pyproject.toml:61]: We really need to get the docs up to spec at some point :(

**Review by richfitz — APPROVED** [review]
Nice - this took me a while to remember how to get everything going, which is itself a problem.  I've been poking around with uv but it's alarming that to get going with this one needs to install uv or pipx to install hatch to install the dependencies, which seems wild.

This worked first time for hitting packit, the oauth flow is nice though not quite as slick as httr2's and I get what looks like some debug printed:

```
Logging in to https://packit.dide.ic.ac.uk/training/
Visit https://github.com/login/device and enter the code <3F72-22EC>
DeviceAuthorizationResponse(device_code='fb462686ed23572ca5f1079c0db99a673391f355', user_code='3F72-22EC', verification_uri='https://github.com/login/device', expires_in=899, interval=5)
Logged in successfully
```

so somewhere within this range: https://github.com/mrc-ide/pyorderly/blob/mrc-5968/src/pyorderly/outpack/location_packit.py#L175-L187 - but it's not obvious what is causing that.

I recognise a lot of the code from the R version, so let's merge this and iterate on the differences.

Up to you how much of the coverage holes you want to fix or ignore

**Pull request merged**